homeassistant/components/hafas/config_flow.py

```diff
@@ -0,0 +1,130 @@
+"""Config flow for HaFAS integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from pyhafas import HafasClient
+from pyhafas.profile import DBProfile, VSNProfile
+from pyhafas.types.fptf import Station
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.backports.enum import StrEnum
+from homeassistant.const import CONF_OFFSET
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import selector
+
+from .const import CONF_DESTINATION, CONF_ONLY_DIRECT, CONF_PROFILE, CONF_START, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Profile(StrEnum):
+    """Enum of HaFAS profile type."""
+
+    DB = "DB"
+    VSN = "VSN"
+
+
+PROFILE_OPTIONS = [
+    selector.SelectOptionDict(value=Profile.DB, label="Deutsche Bahn"),
+    selector.SelectOptionDict(
+        value=Profile.VSN, label="Verkehrsverbund Süd-Niedersachsen"
+    ),
+]
+
+DEFAULT_OFFSET = {"seconds": 0}
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_PROFILE): selector.SelectSelector(
+            selector.SelectSelectorConfig(
+                options=PROFILE_OPTIONS,
+                multiple=False,
+                mode=selector.SelectSelectorMode.DROPDOWN,
+            ),
+        ),
+        vol.Required(CONF_START): str,
+        vol.Required(CONF_DESTINATION): str,
+        vol.Required(CONF_OFFSET, default=DEFAULT_OFFSET): selector.DurationSelector(),
+        vol.Required(CONF_ONLY_DIRECT, default=False): bool,
+    }
+)
+
+
+def validate_station_name(client: HafasClient, station_name: str) -> Station:
+    """Validate station based on user input."""
+
+    stations = client.locations(station_name)
+    if len(stations) == 0:
+        return None
+
+    return stations[0]
```

Comment:
The API seems to be very easy-going in its matching, so I can imagine people getting a different station they expected from time to time if the first result is always chosen. Maybe only take the first result if there's only one matching station, or when if the search query matches the first result 100%? And in other cases list like the first x matching stations in the error message.

In theory, you could also provide HA with a list of stations to limit the results only to valid options, but I suspect that might be a bit too much to handle in the UI.

Reply:
I didn't like the error approach so I added a second form step, with a dropdown of the suggested start and destination stations.